{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "545005ec-61b3-4b53-8a49-b23b1a402507",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Import dependencies ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8e23ec95-6132-4318-a0ff-910b00239a97",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import time\n",
    "import warnings\n",
    "import math\n",
    "import csv\n",
    "import h5py\n",
    "import string\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.colors as mcolors\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.ticker import LogFormatterMathtext\n",
    "from mpl_scatter_density import ScatterDensityArtist\n",
    "import pyideogram\n",
    "import pybigtools\n",
    "import scipy.optimize\n",
    "import scipy.io\n",
    "from scipy.ndimage import gaussian_filter\n",
    "from scipy.optimize import curve_fit, fsolve\n",
    "from numpy import genfromtxt\n",
    "from typing import Any\n",
    "from time import monotonic\n",
    "import cProfile\n",
    "from random import random\n",
    "from itertools import accumulate\n",
    "from math import floor\n",
    "import gzip\n",
    "from Bio import SeqIO\n",
    "\n",
    "# General options\n",
    "np.set_printoptions(threshold=sys.maxsize)\n",
    "\n",
    "# Suppress specific warnings\n",
    "warnings.filterwarnings(\"ignore\", message=\"The iteration is not making good progress\")\n",
    "warnings.filterwarnings(\"ignore\", message=\".*Creating legend with loc=\\\"best\\\" can be slow with large amounts of data.*\")\n",
    "warnings.filterwarnings(\"ignore\", message=\".*All-NaN slice encountered.*\", category=RuntimeWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4da929f-b18c-44f4-9a69-07476a0b6b80",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### General variables ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6402efde-5206-4a1f-96c2-380778c39811",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_lines = [\"HeLa-S3\",\"BJ1\",\"IMR90\",\"HUVEC\",\"K562\",\"GM12878\",\"HepG2\",\"MCF-7\",\"H1\",\"H9\",\"HCT\"]\n",
    "chr_lengths = [249251, 243200, 198023, 191155, 180916, 171116, 159139, 146365, 141214, 135535, 135007, 133852, 115170, 107350, 102532, 90355, 81196, 78078, 59129, 63026, 48130, 51305]\n",
    "\n",
    "list1 = [\"time_data\", \"time_sim\", \"error\", \"fire_rates\", \"forkd\",\"telomeres\"]\n",
    "list2 = [\"Replication time (min)\", \"Replication time (min)\", \"Error\", \"Firing rate\", \"Fork directionality\", \"Telomeres\"]\n",
    "title_map = {key: value for key, value in zip(list1, list2)}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c701f5c-5717-42c9-9bdd-592baa314ab5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Data generation ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a7d728ea-1786-4f81-aa27-d9836d512292",
   "metadata": {},
   "outputs": [],
   "source": [
    "### BigWig data ###\n",
    "# From: https://genome.ucsc.edu/cgi-bin/hgFileUi?db=hg19&g=wgEncodeUwRepliSeq\n",
    "\n",
    "def sigmoid(values, k):\n",
    "    values = np.array(values)  \n",
    "    if k > 0:\n",
    "        return 50 * (1 + np.tanh((k / 100) * (values - 50)) / np.tanh(0.5 * k))\n",
    "    elif k == 0:\n",
    "        return values  # Identity function when k=0\n",
    "\n",
    "def datagenBigWig(cell_line, chr, minp, maxp, resolution, alld, dtscale, saveQ, info, sigscale=0):\n",
    "    file_path = f'data/bigwig_files/{cell_line}.bigWig'\n",
    "    bw = pybigtools.open(open(file_path, 'rb'))  # Keep the original file opening method as requested\n",
    "    time_data_all = bw.values(f'chr{chr}')\n",
    "\n",
    "    global time_data\n",
    "    \n",
    "    if not alld:\n",
    "        time_data_all = bw.values(f'chr{chr}', minp * resolution, maxp * resolution)\n",
    "    \n",
    "    # Sample equally spaced values from `time_data_all` with the given resolution\n",
    "    time_data = np.array(time_data_all[::resolution])\n",
    "    \n",
    "\n",
    "    # Identify invalid positions\n",
    "    invalid_positions = np.where(np.isnan(time_data) | (time_data <= 0))[0]\n",
    "    \n",
    "    # Filter the time_data\n",
    "    time_data = np.nan_to_num(time_data, nan=0.0001)  # Map 'nan' to 0.0001\n",
    "    time_data[time_data <= 0] = 0.0001  # Map values less than or equal to 0 to 0.0001\n",
    "    time_data[time_data > 100] = 100  # Cap values greater than 100 to 100\n",
    "    time_data = np.array([100 - i for i in time_data]) # Data is given in inversed scale\n",
    "    # Optional: Apply sigmoid transformation (0 for no transform)\n",
    "    #time_data = sigmoid(time_data, sigscale) # Use k = (0,2,5,10,50)\n",
    "    # Scaling\n",
    "    time_data = dtscale * time_data\n",
    "    interval_min = 30 # The start of the S-phase is often detected within a range of 0 to 30 minutes into the S-phase.\n",
    "    interval_max = max(time_data)\n",
    "    time_data = (time_data - np.min(time_data) )/ (np.max(time_data) - np.min(time_data)) * (interval_max - interval_min) + interval_min\n",
    "    time_data[invalid_positions] = max(time_data)\n",
    "\n",
    "    if saveQ:\n",
    "        np.savetxt(f\"data/whole-genome_timing_data/time_data_{info}.txt\", time_data, fmt='%.30f')\n",
    "        np.savetxt(f\"data/whole-genome_missing_data/missing_data_{info}.txt\", invalid_positions, fmt='%d')\n",
    "\n",
    "    return time_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a6206e1a-decb-454f-a3b7-faeffd4e7680",
   "metadata": {},
   "outputs": [],
   "source": [
    "### High-resolution data ###\n",
    "# From: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE137764\n",
    "\n",
    "def logistic(x, L, k, x0):\n",
    "    \"\"\"Logistic function used for curve fitting.\"\"\"\n",
    "    return L / (1 + np.exp(-k * (x - x0)))\n",
    "\n",
    "def calculate_medians(rtil2):\n",
    "    \"\"\"Calculates the median time points for each list of observations.\"\"\"\n",
    "    medians = []\n",
    "\n",
    "    for data in rtil2:\n",
    "        if np.sum(data) == 0:\n",
    "            medians.append(np.nan)  # Handle the case of all zeros or no data\n",
    "            continue\n",
    "\n",
    "        # Accumulate the data points\n",
    "        accumulated_data = np.cumsum(data)\n",
    "\n",
    "        # Normalize the accumulated data to have a final value of 1\n",
    "        normalized_data = accumulated_data / accumulated_data[-1]\n",
    "\n",
    "        # Time points evenly distributed over 10 hours\n",
    "        time_points = np.linspace(0, 10, len(data))\n",
    "\n",
    "        # Fit the logistic function to the normalized accumulated data\n",
    "        try:\n",
    "            params, _ = curve_fit(logistic, time_points, normalized_data, p0=[1, 1, 5])\n",
    "            L, k, x0 = params\n",
    "            medians.append(x0)  # Append the median time point\n",
    "        except RuntimeError:\n",
    "            medians.append(np.nan)  # Append NaN if the fit fails\n",
    "\n",
    "    return medians\n",
    "\n",
    "def datagenHighRes(cell_line, chr, minp, maxp, resolution, alld, dtscale, saveQ, info):\n",
    "\n",
    "    global time_data\n",
    "    \n",
    "    # Lengths of each chromosome in kilobases\n",
    "    chr_lengths = [249251, 243200, 198023, 191155, 180916, 171116, 159139, 146365, 141214, 135535, 135007, 133852, 115170, 107350, 102532, 90355, 81196, 78078, 59129, 63026, 48130, 51305]\n",
    "\n",
    "    # Path to the data file\n",
    "    matfile = f'data/high_res_files/GSE137764_{cell_line}_GaussiansGSE137764_mooth_scaled_autosome.mat'\n",
    "    \n",
    "    # Read the data\n",
    "    data = pd.read_csv(matfile, delimiter=\"\\t\", low_memory=False)\n",
    "    \n",
    "    # Extract relevant columns for the chromosome\n",
    "    selected_columns = [col for col in data.columns if (f'chr{str(chr)}' == col or f'chr{str(chr)}.' in col)]\n",
    "    \n",
    "    rtil1 = []\n",
    "    for col in selected_columns:\n",
    "        lcol = np.array(data[col][2:18])\n",
    "        lcol[np.isnan(lcol)] = 0.  # Ensures no NaNs are processed\n",
    "        rtil1.append(lcol)\n",
    "\n",
    "    rtil2 = calculate_medians(rtil1)  # Assume this function calculates some form of median or summarization\n",
    "    \n",
    "    # Calculate repeat factor and ensure the length is exactly the chromosome length\n",
    "    original_length = len(rtil2)\n",
    "    repeat_factor = chr_lengths[chr - 1] // original_length + (chr_lengths[chr - 1] % original_length > 0)\n",
    "    \n",
    "    # Create the repeated array, then slice to the exact chromosome length\n",
    "    extended_data = np.repeat(rtil2, repeat_factor)[:chr_lengths[chr - 1]]\n",
    "    \n",
    "    # Apply Gaussian smoothing\n",
    "    sigma = 20  # Standard deviation for Gaussian smoothing\n",
    "    time_data = gaussian_filter(extended_data, sigma=sigma)\n",
    "\n",
    "    invalid_positions = np.where(np.isnan(time_data) | (time_data <= 0))[0]\n",
    "    \n",
    "    # Ensure there are no NaN values in the final output\n",
    "    time_data = np.nan_to_num(time_data)\n",
    "    time_data = 60 * time_data\n",
    "    time_data[invalid_positions] = max(time_data)\n",
    "\n",
    "    if saveQ:\n",
    "        np.savetxt(f\"data/whole-genome_timing_data/time_data_{info}.txt\", time_data, fmt='%.30f')\n",
    "        np.savetxt(f\"data/whole-genome_missing_data/missing_data_{info}.txt\", invalid_positions, fmt='%d')\n",
    "    \n",
    "    return time_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a52a5644-ad0c-4bb9-88fa-fe4f80e9ec35",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Simple data generation () ###\n",
    "# To be used in fitting\n",
    "def datagenfs(cell_line, chr_number, chrpos_min, chrpos_max, resolution, alld, dtscale, saveQ, info, sigscale=0):\n",
    "    if alld:\n",
    "        time_data = np.loadtxt(f'data/whole-genome_timing_data/time_data_{cell_line}_chr[{chr_number}].txt', dtype=float)\n",
    "    else:\n",
    "        time_data = np.loadtxt(f'data/whole-genome_timing_data/time_data_{cell_line}_chr[{chr_number}].txt', dtype=float)[chrpos_min:chrpos_max]\n",
    "        np.savetxt(f\"data/whole-genome_timing_data/time_data_{cell_line}_chr[{chr_number}]_{chrpos_min}-{chrpos_max}.txt\", time_data, fmt='%.30f')\n",
    "    return time_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab490c85-f1da-4a03-ab60-01946c48a29d",
   "metadata": {},
   "source": [
    "#### GRO-Seq files ####"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d04a065a-64fe-43c4-a3ff-4fb074f633d0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Fitting ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "55c3ba13-3447-47a7-a1ff-36f7428263d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fitfunction(list, v0, st0, fit_step, maxiter, err_threshold, saveQ, info):\n",
    "    \n",
    "    timel = list\n",
    "    \n",
    "    v = v0\n",
    "    st = st0\n",
    "    exp_v = np.exp(-1/v)\n",
    "    x00 = np.array([(math.pi/(4*v))*i**(-2) for i in timel])\n",
    "    lm = 1000 # Remove end regions for error calculation\n",
    "    \n",
    "    # VECTORIZED APPROACH\n",
    "    \n",
    "    def mse(y_true, y_pred):\n",
    "        mse_value = sum((yt - yp) ** 2 for yt, yp in zip(y_true, y_pred)) / len(y_true)\n",
    "        return mse_value\n",
    "    \n",
    "    def fast_roll_add(dst, src, shift):\n",
    "        dst[shift:] += src[:-shift]\n",
    "        dst[:shift] += src[-shift:]\n",
    "    \n",
    "    # Expected replication time computation (replaces bcs)\n",
    "    def fp(x, L, v):\n",
    "        n = len(x)\n",
    "        y = np.zeros(n)\n",
    "    \n",
    "        last_exp_2_raw = np.zeros(n)\n",
    "        last_exp_2 = np.ones(n)\n",
    "        unitary = x.copy()\n",
    "        for k in range(L+1):\n",
    "            if k != 0:\n",
    "                fast_roll_add(unitary, x, k)\n",
    "                fast_roll_add(unitary, x, -k)\n",
    "            exp_1_raw = last_exp_2_raw\n",
    "            exp_1 = last_exp_2\n",
    "            exp_2_raw = exp_1_raw + unitary / v\n",
    "            exp_2 = np.exp(-exp_2_raw)\n",
    "    \n",
    "            # Compute the weighted sum for each j and add to the total\n",
    "            y += (exp_1 - exp_2) / unitary\n",
    "            \n",
    "            last_exp_2_raw = exp_2_raw\n",
    "            last_exp_2 = exp_2\n",
    "        return y\n",
    "\n",
    "    # Fitting iteration\n",
    "    def fitf(time, lst, x0, j, fit_step):\n",
    "        return x0[j] * (lst[j] / time[j])**(fit_step)\n",
    "\n",
    "    # Alternative fitting\n",
    "    def fitf0(time, lst, x0, j, fit_step):\n",
    "        return x0[j]**(np.log(time[j]) / np.log(lst[j]))\n",
    "\n",
    "    # Fitting control\n",
    "    def cfit(time, lst, x0, fit_step):\n",
    "        result = np.empty_like(x0)\n",
    "        for j in range(len(x0)):\n",
    "            fit_result = fitf(time, lst, x0, j, fit_step)\n",
    "            if fit_result < 10**(-err_threshold):\n",
    "                result[j] = 10**(-err_threshold)\n",
    "            #elif abs(time[j] - lst[j]) < .5:\n",
    "            #    result[j] = x0[j]\n",
    "            else:\n",
    "                result[j] = fit_result\n",
    "        return result\n",
    "    \n",
    "    xs = x00\n",
    "    ys = fp(xs, len(xs)//st, v)\n",
    "    err = 10**10\n",
    "    \n",
    "    for j in range(maxiter):\n",
    "        xs0 = xs\n",
    "        ys0 = ys\n",
    "        xs = cfit(timel, ys, xs, fit_step)\n",
    "        ys = fp(xs, len(xs)//st, v)\n",
    "        \n",
    "        new_err = mse(timel[lm:-lm], ys[lm:-lm])\n",
    "        print(str(j+1) + '/' + str(maxiter) + ' err: ' + str('{:.30f}'.format(new_err)), end=\"\\r\")\n",
    "        \n",
    "        err = new_err  # Update the error with the new calculated error\n",
    "\n",
    "    fire_rates = ['{:.30f}'.format(i) for i in xs]\n",
    "    time_sim = ys\n",
    "    \n",
    "    if saveQ:\n",
    "        with open(r'data/whole-genome_firing_rates/fire_rates_'+info+'.txt', 'w') as f:\n",
    "            for rate in fire_rates:\n",
    "                f.write(rate + '\\n')\n",
    "        np.savetxt(r'data/whole-genome_timing_simulation/time_sim_'+info+'.txt', time_sim, fmt='%.30f')\n",
    "    \n",
    "    return [fire_rates, time_sim]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d98edf98-ea6d-4f6d-aaf4-f69b89c2d44a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Error generation ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e2994092-64b9-4128-812d-722816580b62",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_squared_error(time_data, time_simulation):\n",
    "    return (time_data - time_simulation) ** 2\n",
    "\n",
    "def process_files_and_compute_squared_error(cell_lines, chr_numbers, base_path):\n",
    "    for cell_line in cell_lines:\n",
    "        for chr_number in chr_numbers:\n",
    "            # Define file paths\n",
    "            time_data_file = os.path.join(base_path, f'whole-genome_timing_data/time_data_{cell_line}_chr[{chr_number}].txt')\n",
    "            time_simulation_file = os.path.join(base_path, f'whole-genome_timing_simulation/time_sim_{cell_line}_chr[{chr_number}].txt')\n",
    "            error_file = os.path.join(base_path, f'whole-genome_error/error_{cell_line}_chr[{chr_number}].txt')\n",
    "\n",
    "            # Load data\n",
    "            time_data = np.loadtxt(time_data_file, dtype=float)\n",
    "            time_simulation = np.loadtxt(time_simulation_file, dtype=float)\n",
    "\n",
    "            # Compute squared error\n",
    "            squared_error = compute_squared_error(time_data, time_simulation)\n",
    "            \n",
    "            # Save squared error to file\n",
    "            np.savetxt(error_file, squared_error, fmt='%.30f')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50fb7139-d379-4755-bfbf-e43b74e84a3b",
   "metadata": {},
   "source": [
    "#### bedgraph file generation ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb547147-b0fd-4a61-b9fa-0b5b881d837d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dictionary of chromosome sizes for hg18\n",
    "def chr_size_fun(genome_build):\n",
    "    if genome_build == 'hg18':\n",
    "        return {\n",
    "            '1': 247249719, '2': 242951149, '3': 199501827, '4': 191273063, '5': 180857866, \n",
    "            '6': 170899992, '7': 158821424, '8': 146274826, '9': 140273252, '10': 135374737, \n",
    "            '11': 134452384, '12': 132349534, '13': 114142980, '14': 106368585, '15': 100338915, \n",
    "            '16': 88827254, '17': 78774742, '18': 76117153, '19': 63811651, '20': 62435964, \n",
    "            '21': 46944323, '22': 49691432, 'X': 154913754, 'Y': 57772954\n",
    "        }\n",
    "    elif genome_build == 'hg19':\n",
    "        return {\n",
    "            '1': 249250621, '2': 243199373, '3': 198022430, '4': 191154276, '5': 180915260,\n",
    "            '6': 171115067, '7': 159138663, '8': 146364022, '9': 141213431, '10': 135534747,\n",
    "            '11': 135006516, '12': 133851895, '13': 115169878, '14': 107349540, '15': 102531392,\n",
    "            '16': 90354753, '17': 81195210, '18': 78077248, '19': 59128983, '20': 63025520,\n",
    "            '21': 48129895, '22': 51304566, 'X': 155270560, 'Y': 59373566\n",
    "        }\n",
    "    elif genome_build == 'hg38':\n",
    "        return {\n",
    "            '1': 248956422, '2': 242193529, '3': 198295559, '4': 190214555, '5': 181538259,\n",
    "            '6': 170805979, '7': 159345973, '8': 145138636, '9': 138394717, '10': 133797422,\n",
    "            '11': 135086622, '12': 133275309, '13': 114364328, '14': 107043718, '15': 101991189,\n",
    "            '16': 90338345, '17': 83257441, '18': 80373285, '19': 58617616, '20': 64444167,\n",
    "            '21': 46709983, '22': 50818468, 'X': 156040895, 'Y': 57227415\n",
    "        }\n",
    "\n",
    "def txt_to_bedgraph(cell_line, data_type='error', genome_build='hg19'):\n",
    "    # Define the output file name\n",
    "    if data_type=='error':\n",
    "        output_file = f'data/whole-genome_error/bedgraph_files/error_{cell_line}.bedgraph'\n",
    "    elif data_type=='fire_rates':\n",
    "        output_file = f'data/whole-genome_firing_rates/bedgraph_files/fire_rates_{cell_line}.bedgraph'\n",
    "    \n",
    "    # Open the output file in write mode\n",
    "    with open(output_file, 'w') as bedgraph:\n",
    "        # Write the header line\n",
    "        bedgraph.write('track type=bedGraph\\n')\n",
    "        \n",
    "        # Iterate through all the txt files for the specified cell line\n",
    "        for chr_number in list(map(str, range(1, 23))) + ['X', 'Y']:\n",
    "            if data_type=='error':\n",
    "                input_file = f'data/whole-genome_error/error_{cell_line}_chr[{chr_number}].txt'\n",
    "            elif data_type=='fire_rates':\n",
    "                input_file = f'data/whole-genome_firing_rates/fire_rates_{cell_line}_chr[{chr_number}].txt'\n",
    "            chr_size = chr_size_fun(genome_build)[chr_number]\n",
    "            \n",
    "            # Check if the input file exists\n",
    "            if os.path.isfile(input_file):\n",
    "                with open(input_file, 'r') as infile:\n",
    "                    # Read through each line and write the corresponding bedgraph entry\n",
    "                    for position, value in enumerate(infile):\n",
    "                        start = position * 1000\n",
    "                        end = min(start + 1000, chr_size)\n",
    "                        if start >= chr_size:\n",
    "                            break\n",
    "                        value = value.strip()\n",
    "                        bedgraph.write(f'chr{chr_number}\\t{start}\\t{end}\\t{value}\\n')\n",
    "            else:\n",
    "                None\n",
    "                \n",
    "    print(f'BEDGRAPH file created: {output_file}', end=\"\\r\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "254806eb-f568-4a39-85f9-c03dbcba5830",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### BCS file generation ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8f436032-2f6c-4b23-a315-ab34e7703950",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bcs_gen(cell_line, chr_number, chrpos_min, chrpos_max, fork_speed, fire_rates, resolution):\n",
    "\n",
    "    file_name = cell_line+'_chr['+str(chr_number)+']_'+str(chrpos_min)+'-'+str(chrpos_max)\n",
    "    \n",
    "    bcfile = 'code/DNAReplication.bc'\n",
    "    new_bcfile = f'code/bcs_scripts/DNAReplication_{cell_line}_chr[{chr_number}]_{chrpos_min}-{chrpos_max}.bc'\n",
    "    bcsfile = []\n",
    "    \n",
    "    chrLength = chrpos_max - chrpos_min\n",
    "    orign = int(chrLength * resolution / 1000)\n",
    "    fast = 100000\n",
    "    x = np.linspace(chrpos_min, chrpos_max, chrLength)  # Chromosome positions\n",
    "    \n",
    "    with open(bcfile, 'r') as file:\n",
    "        bcsfile = file.readlines()\n",
    "    bcsfile[bcsfile.index(\"// Chromosome length\\n\")+1] = \"L = \"+str(chrLength)+\";\\n\"\n",
    "    bcsfile[bcsfile.index(\"// Fast rate\\n\")+1] = \"fast = \"+str(fast)+\";\\n\"\n",
    "    bcsfile[bcsfile.index(\"// Fork velocity\\n\")+1] = \"v = \"+str(fork_speed)+\";\\n\"\n",
    "    \n",
    "    oril = list(map(floor, np.linspace(1, chrLength, num=orign)))\n",
    "    \n",
    "    flistn = fire_rates\n",
    "    \n",
    "    # write new origins\n",
    "    oriarr = np.array([\n",
    "        'ORI[' + str(floor(oril[i1])) + ',' + '{:.30f}'.format(flistn[i1]) + ']'\n",
    "        for i1 in range(0, orign)\n",
    "    ])\n",
    "    \n",
    "    # delete all the origins\n",
    "    with open(new_bcfile, 'w') as fp:\n",
    "        for number, line in enumerate(bcsfile):\n",
    "            if number not in range(bcsfile.index(\"// PROCESS INITIATION\\n\")+1, bcsfile.index(\"// END\")-2):\n",
    "                fp.write(line)\n",
    "        \n",
    "    # now change the last line\n",
    "    with open(new_bcfile, 'r') as file:\n",
    "        bcsfile = file.readlines()\n",
    "        bcsfile[bcsfile.index(\"// PROCESS INITIATION\\n\")+1] = str(oriarr).replace('\"','').replace(\"'\",'').replace(\" \",\" || \")[1:-1]+';\\n'\n",
    "    \n",
    "    with open(new_bcfile, 'w') as file:\n",
    "        file.writelines(bcsfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f85fb9f-01aa-4173-b4d0-063f3dabca2e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### BCS simulation output ####"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6cabe8e-2d2e-46e6-a52d-0711f95f2b6e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "##### Replication timing, fork directionality and origins #####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "87aa4633-909b-4825-867c-6bb36db250f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_bcs_output(cell_line, chr_number, chrpos_min, chrpos_max, fork_speed, resolution, scale_factor, sim_number, compute_replication_time, compute_fork_directionality, compute_origin_positions):\n",
    "    # Define the file path\n",
    "    file_path = f'data/bcs_output/bcs_output_{cell_line}_chr[{chr_number}]_{chrpos_min}-{chrpos_max}.simulation.bcs'\n",
    "\n",
    "    # Initialize arrays to store replication time and fork directionality\n",
    "    DNA_replicationtime = [0.0 for _ in range(0, chrpos_max - chrpos_min)] if compute_replication_time else None\n",
    "    DNA_forkdirectionality = [0.0 for _ in range(0, chrpos_max - chrpos_min)] if compute_fork_directionality else None\n",
    "    DNA_originpositions = [] if compute_origin_positions else None  # List to store origin positions per simulation\n",
    "    current_origins = []\n",
    "    sim_iteration = 0\n",
    "\n",
    "    with open(file_path) as f:\n",
    "        for line in f:\n",
    "            if sim_iteration == sim_number + 1:\n",
    "                break\n",
    "            if line[0] == '>':\n",
    "                alreadyDone = []\n",
    "                if compute_origin_positions and current_origins:  # If we have collected origins for the current simulation\n",
    "                    DNA_originpositions.append(current_origins)\n",
    "                current_origins = []\n",
    "                print(sim_iteration, end=\"\\r\")\n",
    "                sim_iteration += 1\n",
    "                continue\n",
    "            splitLine = line.split('\\t')\n",
    "            if compute_origin_positions and splitLine[2] == \"ORI\":\n",
    "                origin_pos = int(splitLine[4])\n",
    "                current_origins.append(origin_pos)\n",
    "            if splitLine[2] == \"FL\":\n",
    "                pos = int(splitLine[4]) - 1\n",
    "                time = float(splitLine[0])\n",
    "                if pos not in alreadyDone:\n",
    "                    if compute_replication_time:\n",
    "                        DNA_replicationtime[pos] += time\n",
    "                    if compute_fork_directionality:\n",
    "                        DNA_forkdirectionality[pos] -= 1  # Track left-moving forks\n",
    "                    alreadyDone.append(pos)\n",
    "            if splitLine[2] == \"FR\":\n",
    "                pos = int(splitLine[4]) - 1\n",
    "                time = float(splitLine[0])\n",
    "                if pos not in alreadyDone:\n",
    "                    if compute_replication_time:\n",
    "                        DNA_replicationtime[pos] += time\n",
    "                    if compute_fork_directionality:\n",
    "                        DNA_forkdirectionality[pos] += 1  # Track right-moving forks\n",
    "                    alreadyDone.append(pos)\n",
    "\n",
    "    # Don't forget to add the origins of the last simulation\n",
    "    if compute_origin_positions and current_origins:\n",
    "        DNA_originpositions.append(current_origins)\n",
    "\n",
    "    # Average the results over the number of simulations\n",
    "    if compute_replication_time:\n",
    "        for i in range(len(DNA_replicationtime)):\n",
    "            DNA_replicationtime[i] = float(DNA_replicationtime[i]) / float(sim_number)\n",
    "\n",
    "    if compute_fork_directionality:\n",
    "        for i in range(len(DNA_forkdirectionality)):\n",
    "            DNA_forkdirectionality[i] = float(DNA_forkdirectionality[i]) / float(sim_number)\n",
    "\n",
    "    # Define file paths for saving the results\n",
    "    base_path = 'data'\n",
    "    replication_time_path = os.path.join(base_path, 'whole-genome_timing_bcs', f'time_bcs_{cell_line}_chr[{chr_number}]_{chrpos_min}-{chrpos_max}.txt')\n",
    "    fork_directionality_path = os.path.join(base_path, 'whole-genome_fork_directionality', f'fork_directionality_{cell_line}_chr[{chr_number}]_{chrpos_min}-{chrpos_max}.txt')\n",
    "    origin_positions_path = os.path.join(base_path, 'whole-genome_origins', f'origin_positions_{cell_line}_chr[{chr_number}]_{chrpos_min}-{chrpos_max}.txt')\n",
    "\n",
    "    # Create directories if they do not exist\n",
    "    os.makedirs(os.path.dirname(replication_time_path), exist_ok=True)\n",
    "    os.makedirs(os.path.dirname(fork_directionality_path), exist_ok=True)\n",
    "    os.makedirs(os.path.dirname(origin_positions_path), exist_ok=True)\n",
    "\n",
    "    # Save the results to text files\n",
    "    if compute_replication_time:\n",
    "        np.savetxt(replication_time_path, DNA_replicationtime, fmt='%.6f')\n",
    "\n",
    "    if compute_fork_directionality:\n",
    "        np.savetxt(fork_directionality_path, DNA_forkdirectionality, fmt='%.6f')\n",
    "\n",
    "    if compute_origin_positions:\n",
    "        with open(origin_positions_path, 'w') as f:\n",
    "            for origins in DNA_originpositions:\n",
    "                f.write(' '.join(map(str, origins)) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96348c62-df12-4924-980d-3ae248b3c245",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_intervals(cell_lines, chr_numbers, fork_speed=1.4, resolution=1000, scale_factor=6, sim_number=5, compute_replication_time=True, compute_fork_directionality=True, compute_origin_positions=True, interval=None):\n",
    "    for cell_line in cell_lines:\n",
    "        for chr_number in chr_numbers:\n",
    "            chr_length = chr_lengths[chr_number - 1]  # Get the length of the chromosome\n",
    "            intervals = [(interval[0], interval[1])] if interval else [(start, min(start + 10000, chr_length)) for start in range(0, chr_length, 10000)]\n",
    "            for start, end in intervals:\n",
    "                process_bcs_output(\n",
    "                    cell_line=cell_line,\n",
    "                    chr_number=chr_number,\n",
    "                    chrpos_min=start,\n",
    "                    chrpos_max=end,\n",
    "                    fork_speed=fork_speed,\n",
    "                    resolution=resolution,\n",
    "                    scale_factor=scale_factor,\n",
    "                    sim_number=sim_number,\n",
    "                    compute_replication_time=compute_replication_time,\n",
    "                    compute_fork_directionality=compute_fork_directionality,\n",
    "                    compute_origin_positions=compute_origin_positions\n",
    "                )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "716f38a9-20e1-4e9e-9662-6698983f39ff",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "##### Interorigin distances #####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "47bb76cd-99d2-44e8-ad5f-0622dcdaaa90",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_interorigin_distances(cell_line, chr_number, chrpos_min, chrpos_max):\n",
    "    base_path = 'data'\n",
    "    \n",
    "    # Define file path for loading the origins\n",
    "    origins_path = os.path.join(base_path, 'whole-genome_origins', f'origin_positions_{cell_line}_chr[{chr_number}]_{chrpos_min}-{chrpos_max}.txt')\n",
    "    \n",
    "    # Load origins data from text file\n",
    "    if os.path.exists(origins_path):\n",
    "        with open(origins_path, 'r') as f:\n",
    "            origins_data = [list(map(int, line.strip().strip('[]').split())) for line in f]\n",
    "    else:\n",
    "        raise FileNotFoundError(f\"Origins data not found at {origins_path}\")\n",
    "    \n",
    "    # Compute interorigin distances for each simulation\n",
    "    interorigin_distances = []\n",
    "    for origins in origins_data:\n",
    "        origins_sorted = sorted(origins)\n",
    "        distances = np.diff(origins_sorted)\n",
    "        interorigin_distances.append(distances)\n",
    "    \n",
    "    # Define file path for saving the interorigin distances\n",
    "    iod_path = os.path.join(base_path, 'whole-genome_interorigin_distances', f'iod_{cell_line}_chr[{chr_number}]_{chrpos_min}-{chrpos_max}.txt')\n",
    "    \n",
    "    # Save interorigin distances to a text file\n",
    "    with open(iod_path, 'w') as f:\n",
    "        for distances in interorigin_distances:\n",
    "            f.write(f\"{list(distances)}\\n\")\n",
    "\n",
    "def compute_interorigin_intervals(cell_lines, chr_numbers, interval=None):\n",
    "    for cell_line in cell_lines:\n",
    "        for chr_number in chr_numbers:\n",
    "            chr_length = chr_lengths[chr_number - 1]  # Get the length of the chromosome\n",
    "            intervals = [(interval[0], interval[1])] if interval else [(start, min(start + 10000, chr_length)) for start in range(0, chr_length, 10000)]\n",
    "            for start, end in intervals:\n",
    "                compute_interorigin_distances(\n",
    "                    cell_line=cell_line,\n",
    "                    chr_number=chr_number,\n",
    "                    chrpos_min=start,\n",
    "                    chrpos_max=end\n",
    "                )\n",
    "\n",
    "def average_iod_data(cell_lines, chr_numbers, factor_min=5, show_per_cell_line=False):\n",
    "    all_iod_data = []\n",
    "    iod_data_per_cell_line = []\n",
    "\n",
    "    for cell_line in cell_lines:\n",
    "        cell_line_iod_data = []\n",
    "        for chr_number in chr_numbers:\n",
    "            chr_length = chr_lengths[chr_number - 1]\n",
    "            for start in range(0, chr_length, 10000):\n",
    "                end = min(start + 10000, chr_length)\n",
    "                iod_data = load_function_metrics(cell_line, chr_number, \"iod\", start, end, factor_min=factor_min)\n",
    "                cell_line_iod_data.extend(iod_data)\n",
    "        iod_data_per_cell_line.append(cell_line_iod_data)\n",
    "        all_iod_data.extend(cell_line_iod_data)\n",
    "\n",
    "    if show_per_cell_line:\n",
    "        return iod_data_per_cell_line\n",
    "    else:\n",
    "        return [all_iod_data]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3770db5-616c-45eb-bdcc-b3394e7f800b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### File joining functions ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e3a807da-65ff-4919-bc1c-a920a0af5eba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def join_files(cell_line, chr_number, datatype, interval=10000):\n",
    "    all_data = []\n",
    "    max_length = chr_lengths[chr_number - 1]\n",
    "    \n",
    "    for start in range(0, max_length, interval):\n",
    "        end = min(start + interval, max_length)  # Ensure the end does not exceed max_length\n",
    "        file_name = f'data/whole-genome_{datatype}/fork_directionality_{cell_line}_chr[{chr_number}]_{start}-{end}.txt'\n",
    "        \n",
    "        if os.path.exists(file_name):\n",
    "            data = np.loadtxt(file_name, dtype=float)\n",
    "            all_data.append(data)\n",
    "        else:\n",
    "            print(f'Warning: {file_name} does not exist and will be skipped.')\n",
    "    \n",
    "    # Concatenate all data into a single array\n",
    "    if all_data:\n",
    "        concatenated_data = np.concatenate(all_data)\n",
    "    else:\n",
    "        concatenated_data = np.array([])\n",
    "\n",
    "    # Save concatenated data to a new file\n",
    "    output_file = f'data/whole-genome_{datatype}/fork_directionality_{cell_line}_chr[{chr_number}].txt'\n",
    "    np.savetxt(output_file, concatenated_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acf3008c-d9bd-49d2-9b2f-972e94eff663",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Loading functions ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "c2ee3660-7fa6-4e62-b731-393edd0ea8eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_function(cell_line, chr_number, load_type, replace_missingQ=True):\n",
    "    if load_type == 'time_data':\n",
    "        file_path = f'data/whole-genome_timing_data/time_data_{cell_line}_chr[{chr_number}].txt'\n",
    "    elif load_type == 'time_sim':\n",
    "        file_path = f'data/whole-genome_timing_simulation/time_sim_{cell_line}_chr[{chr_number}].txt'\n",
    "    elif load_type == 'error':\n",
    "        file_path = f'data/whole-genome_error/error_{cell_line}_chr[{chr_number}].txt'\n",
    "    elif load_type == 'fire_rates':\n",
    "        file_path = f'data/whole-genome_firing_rates/fire_rates_{cell_line}_chr[{chr_number}].txt'\n",
    "    elif load_type == 'forkd':\n",
    "        file_path = f'data/whole-genome_fork_directionality/fork_directionality_{cell_line}_chr[{chr_number}].txt'\n",
    "        \n",
    "    if not os.path.exists(file_path):\n",
    "        return np.array([], dtype=int)\n",
    "    data = np.loadtxt(file_path, dtype=float)\n",
    "    if replace_missingQ:\n",
    "        missing_data_path = f'data/whole-genome_missing_data/missing_data_{cell_line}_chr[{chr_number}].txt'\n",
    "        if os.path.getsize(missing_data_path) > 0:\n",
    "            missing_positions = np.loadtxt(missing_data_path, dtype=int)\n",
    "        else:\n",
    "            missing_positions = np.array([], dtype=int) \n",
    "        data[missing_positions] = np.nan\n",
    "    return data\n",
    "\n",
    "def load_function_pos(chr_number, load_type, cell_line=None, site_letter='A', base='A'):\n",
    "    if load_type == 'centromeres':\n",
    "        file_path = f'data/genome_regions/centromeres/positions_centromeres_chr[{chr_number}].txt'\n",
    "    elif load_type == 'telomeres':\n",
    "        file_path = f'data/genome_regions/telomeres/positions_telomeres_chr[{chr_number}].txt'\n",
    "    elif load_type == 'fragile_sites':\n",
    "        file_path = f'data/genome_regions/fragile_sites/positions_fragile_site_{chr_number}{site_letter}.txt'\n",
    "    elif load_type == 'bases':\n",
    "        file_path = f'data/genome_regions/bases/positions_{base}_chr[{chr_number}].txt'\n",
    "        \n",
    "    if not os.path.exists(file_path):\n",
    "        return np.array([], dtype=int)\n",
    "    data = np.loadtxt(file_path, dtype=int)\n",
    "    return data\n",
    "\n",
    "def load_function_metrics(cell_line, chr_number, load_type, chrpos_min, chrpos_max, factor_min=5, replace_missingQ=True):\n",
    "    if load_type == \"iod\":\n",
    "        file_path = f'data/whole-genome_interorigin_distances/iod_{cell_line}_chr[{chr_number}]_{chrpos_min}-{chrpos_max}.txt'\n",
    "        iod_data = []\n",
    "        with open(file_path, 'r') as file:\n",
    "            for line in file:\n",
    "                iod_values = list(map(float, line.strip().strip('[]').split(',')))\n",
    "                iod_data.extend([iod for iod in iod_values if iod >= factor_min])  # Filter IOD values\n",
    "        data = np.array(iod_data)\n",
    "        return data\n",
    "\n",
    "def load_missing_data(cell_line, chr_number):\n",
    "    file_path = f'data/whole-genome_missing_data/missing_data_{cell_line}_chr[{chr_number}].txt'\n",
    "    if os.path.exists(file_path) and os.path.getsize(file_path) > 0:\n",
    "        data = np.loadtxt(file_path, dtype=int)\n",
    "    else:\n",
    "        data = np.array([], dtype=int)\n",
    "    return data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e274fa0f-72f0-41a9-a1fe-1c826115bf8f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Genome regions ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4eac2454-c172-403d-aaec-151aecb6c1c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_all_data(cell_line, chr_numbers, load_type1, load_type2):\n",
    "    all_data1 = []\n",
    "    all_data2 = []\n",
    "    for chr_number in chr_numbers:\n",
    "        data1 = load_function(cell_line, chr_number, load_type1)\n",
    "        data2 = load_function(cell_line, chr_number, load_type2)\n",
    "        all_data1.extend(data1)\n",
    "        all_data2.extend(data2)\n",
    "    return [all_data1, all_data2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "87572b19-ee3d-4173-bc81-f2a6e32d62a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Centromeres and telomeres\n",
    "\n",
    "def gen_positions_centromere_telomeres(chr_number):\n",
    "    # Telomere positions (start and end 500 kb)\n",
    "    telomere_start = 500  # in kb\n",
    "    telomere_end_offset = 500  # in kb\n",
    "\n",
    "    # Centromere positions (in kb, hg38, approximate)\n",
    "    centromere_positions_hg38 = [\n",
    "        (121535, 124535), (92326, 95326), (90505, 93505), (49660, 52660),\n",
    "        (46406, 49406), (58830, 61830), (58054, 61054), (43839, 46839),\n",
    "        (47368, 50368), (39255, 42255), (51644, 54644), (34857, 37857),\n",
    "        (16000, 19000), (16000, 19000), (17000, 20000), (35336, 38336),\n",
    "        (22263, 25263), (15461, 18461), (24682, 27682), (26370, 29370),\n",
    "        (11288, 14288), (13000, 16000)\n",
    "    ]\n",
    "\n",
    "    # Provided chromosome lengths (in kb)\n",
    "    chromosome_lengths = [\n",
    "        249251, 243200, 198023, 191155, 180916, 171116, 159139, 146365,\n",
    "        141214, 135535, 135007, 133852, 115170, 107350, 102532, 90355,\n",
    "        81196, 78078, 59129, 63026, 48130, 51305\n",
    "    ]\n",
    "\n",
    "    # Define additional positions to include telomeres and centromeres\n",
    "    length = chromosome_lengths[chr_number - 1]\n",
    "    centromere_start, centromere_end = centromere_positions_hg38[chr_number - 1]\n",
    "    positions_telomeres = np.concatenate([\n",
    "        np.arange(0, telomere_start),\n",
    "        np.arange(length - telomere_end_offset, length)\n",
    "    ])\n",
    "    positions_centromeres = np.arange(centromere_start, centromere_end)\n",
    "\n",
    "    # Save telomere_positions to a text file\n",
    "    np.savetxt(f'data/genome_regions/telomeres/positions_telomeres_chr[{chr_number}].txt', positions_telomeres, fmt='%d')\n",
    "\n",
    "    # Save centromere_positions to a text file\n",
    "    np.savetxt(f'data/genome_regions/centromeres/positions_centromeres_chr[{chr_number}].txt', positions_centromeres, fmt='%d')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "04f3555a-d3c5-48ea-b9d0-fcf1327e4178",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fragile sites\n",
    "\n",
    "def gen_positions_fragile_sites(chr_number, site_letter):\n",
    "    # Load the CSV file\n",
    "    csv_path = 'data/fragile_sites/humCFS-fragile_sites.csv'\n",
    "    df = pd.read_csv(csv_path, header=None)\n",
    "    \n",
    "    positions_fragile_site = []\n",
    "\n",
    "    # Find the column corresponding to the given chromosome\n",
    "    col_index = chr_number - 1  # Chromosome 1 corresponds to column 0, and so on\n",
    "\n",
    "    if col_index >= df.shape[1]:\n",
    "        print(f\"Warning: Chromosome {chr_number} not found in the CSV file.\")\n",
    "        return\n",
    "\n",
    "    # Find the row corresponding to the given site letter\n",
    "    row_index = ord(site_letter.upper()) - ord('A')  # 'A' corresponds to row 0, 'B' to row 1, and so on\n",
    "\n",
    "    if row_index >= df.shape[0]:\n",
    "        #print(f\"Warning: Site letter {site_letter} not found in the CSV file for chromosome {chr_number}.\")\n",
    "        return\n",
    "\n",
    "    # Extract the range in the form chrposmin-chrposmax\n",
    "    site_range = df.iloc[row_index, col_index]\n",
    "\n",
    "    if pd.isna(site_range):\n",
    "        #print(f\"Warning: No data for site {site_letter} on chromosome {chr_number}.\")\n",
    "        return\n",
    "\n",
    "    # Split the range into minimum and maximum positions\n",
    "    pos_min, pos_max = map(int, site_range.split('-'))\n",
    "\n",
    "    # Convert positions to kb\n",
    "    pos_min_kb = pos_min // 1000\n",
    "    pos_max_kb = pos_max // 1000\n",
    "\n",
    "    # Append the range as a numpy array\n",
    "    positions_fragile_site.append(np.arange(pos_min_kb, pos_max_kb))\n",
    "\n",
    "    # Flatten the list of arrays\n",
    "    flattened_positions = np.concatenate(positions_fragile_site)\n",
    "\n",
    "    # Save positions_fragile_site to a text file, with each value on a new line\n",
    "    np.savetxt(f'data/genome_regions/fragile_sites/positions_fragile_site_{chr_number}{site_letter}.txt', flattened_positions, fmt='%d')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2484275d-0f9a-4aa5-a221-f1f331b4f283",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Base regions\n",
    "\n",
    "def gen_positions_bases(local_genome_file, chr_lengths):\n",
    "    bases = ['A', 'T', 'G', 'C']\n",
    "    \n",
    "    output_dir = 'data/genome_regions/bases'\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    \n",
    "    with gzip.open(local_genome_file, \"rt\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"fasta\"):\n",
    "            chr_number = record.id.lstrip(\"chr\")  # Removing 'chr' prefix\n",
    "            if chr_number.isdigit():\n",
    "                chr_number = int(chr_number)\n",
    "                if chr_number in range(1, 23):  # Assuming we only want chromosomes 1-22\n",
    "                    seq = str(record.seq).upper()  # Ensure sequence is uppercase\n",
    "                    base_files = {base: [] for base in bases}\n",
    "                    \n",
    "                    for kb in range(chr_lengths[chr_number-1]):\n",
    "                        position = kb * 1000  # 0-based position\n",
    "                        if position < len(seq):\n",
    "                            base_pair = seq[position]\n",
    "                            if base_pair in base_files:\n",
    "                                base_files[base_pair].append(kb)\n",
    "                    \n",
    "                    for base, locations in base_files.items():\n",
    "                        with open(os.path.join(output_dir, f'positions_{base}_chr[{chr_number}].txt'), 'w') as file:\n",
    "                            for loc in locations:\n",
    "                                file.write(f'{loc}\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67e10b33-356b-4207-9d07-0ac14e18522d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Replication timing plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "1a49d406-3913-45b3-9a5d-b15b8ad1f31f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rt_plotf(cell_line, chr_number, chrpos_min, chrpos_max, scale_factor, file_name, spec_fileQ, saveQ, ax=None, show_ticks=True, show_title=True, simQ=False):\n",
    "    global time_data\n",
    "\n",
    "    # Data loading (Warning: requires saving data in fitting procedure)\n",
    "    # Choose between whole-genome files or particular simulation\n",
    "    if spec_fileQ:\n",
    "        time_data = np.loadtxt(f'data/whole-genome_timing_data/time_data_{cell_line}_chr[{chr_number}]_{chrpos_min}-{chrpos_max}.txt', dtype=float)\n",
    "        time_sim = np.loadtxt(f'data/whole-genome_timing_simulation/time_sim_{cell_line}_chr[{chr_number}]_{chrpos_min}-{chrpos_max}.txt', dtype=float)\n",
    "    else:\n",
    "        time_data = np.loadtxt(f'data/whole-genome_timing_data/time_data_{cell_line}_chr[{chr_number}].txt', dtype=float)[chrpos_min:chrpos_max]\n",
    "        time_sim = np.loadtxt(f'data/whole-genome_timing_simulation/time_sim_{cell_line}_chr[{chr_number}].txt', dtype=float)[chrpos_min:chrpos_max]\n",
    "\n",
    "    if simQ:\n",
    "        time_sim = np.loadtxt(f'data/whole-genome_timing_bcs/time_bcs_{cell_line}_chr[{chr_number}]_{chrpos_min}-{chrpos_max}.txt', dtype=float)\n",
    "    \n",
    "    x = np.linspace(chrpos_min, chrpos_max, chrpos_max - chrpos_min)  # Chromosome positions\n",
    "\n",
    "    \n",
    "    # Plotting\n",
    "    if ax is None:\n",
    "        plt.figure(figsize=(10, 6))\n",
    "        ax = plt.gca()\n",
    "        \n",
    "    ax.plot(x, time_data, label='data', color='gray', linewidth=4, alpha=0.6)\n",
    "    ax.plot(x, time_sim, label='bcs', color='red', linewidth=4, alpha=0.6)\n",
    "    if show_title:\n",
    "        ax.set_title(cell_line + ' - Chromosome ' + str(chr_number))\n",
    "    ax.set_xlabel('Chromosome position (kb)' if show_title else None)\n",
    "    ax.set_ylabel('Time in S-phase (min)' if show_title else \"Replication time\")\n",
    "    ax.set_ylim(100 * scale_factor, 0)\n",
    "    ax.set_xlim(chrpos_min, chrpos_max)  # Ensure the x-axis covers the full range\n",
    "    ax.legend(loc='lower right')\n",
    "    ax.grid(True)\n",
    "    ax.grid(False)\n",
    "    ax.tick_params(axis='both', which='both', direction='out', bottom=show_ticks, labelbottom=show_ticks, left=show_ticks, labelleft=show_ticks)\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_visible(True)\n",
    "    \n",
    "    # Save plot\n",
    "    if saveQ:\n",
    "        plt.savefig('figures/plot_RT_' + file_name + '.pdf', bbox_inches='tight', transparent=True)\n",
    "\n",
    "    if ax is None:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f23b0512-31da-41a6-84ac-097d757fbdf0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### KDE plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "cf735135-fd34-4147-b0bd-c65fffcfcdf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_relative_kdes(data_list, labels, bw_adjust=1, saveQ=False, x_grid_size=1000, normalize=False, log_scale=False, x_min=None, x_max=None, plot_title=\"Relative density plots\", x_title=\"Error\", save_name=\"savedfile\"):\n",
    "    if log_scale:\n",
    "        # Filter out non-positive values for log scale\n",
    "        data_list = [data[data > 0] for data in data_list]\n",
    "        if x_min is None:\n",
    "            x_min = min(data.min() for data in data_list)\n",
    "        if x_max is None:\n",
    "            x_max = max(data.max() for data in data_list)\n",
    "        x = np.logspace(np.log10(x_min), np.log10(x_max), x_grid_size)\n",
    "    else:\n",
    "        if x_min is None:\n",
    "            x_min = min(data.min() for data in data_list)\n",
    "        if x_max is None:\n",
    "            x_max = max(data.max() for data in data_list)\n",
    "        x = np.linspace(x_min, x_max, x_grid_size)\n",
    "\n",
    "    plt.figure(figsize=(6, 6))\n",
    "\n",
    "    # Compute and plot the KDEs\n",
    "    if normalize:\n",
    "        for data in data_list:\n",
    "            ax = sns.kdeplot(data, fill=True, bw_adjust=bw_adjust, log_scale=log_scale)#, alpha=.5)\n",
    "    else:\n",
    "        ax = sns.kdeplot(data_list, fill=True, bw_adjust=bw_adjust, log_scale=log_scale)#, alpha=.5)\n",
    "    handles = [mpatches.Patch(facecolor=color, label=label, alpha=0.5) for color, label in zip(plt.rcParams['axes.prop_cycle'].by_key()['color'], labels)]\n",
    "\n",
    "\n",
    "    \n",
    "    plt.title(plot_title)\n",
    "    plt.xlabel(x_title)\n",
    "    \n",
    "    plt.gca().yaxis.set_visible(False)  # Remove y-axis ticks\n",
    "    plt.ylabel('')  # Remove y-axis label\n",
    "    if log_scale:\n",
    "        plt.xscale('log')\n",
    "    \n",
    "    plt.xlim(x_min, x_max)\n",
    "    plt.legend(handles=handles)\n",
    "    if saveQ:\n",
    "        plt.savefig(f'figures/fig_kdeplot_{save_name}.svg', bbox_inches='tight', transparent=True)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7315a2d0-2517-4d74-b7d4-88b230084463",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Data vs data scatter plots ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "aa6e2754-f6fb-430d-81cc-3110c4eda6af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_modified_cmap(map_to_white):\n",
    "    viridis = plt.cm.viridis\n",
    "    newcolors = viridis(np.linspace(0, 1, 256))\n",
    "    if map_to_white:\n",
    "        white = np.array([1, 1, 1, 1])\n",
    "        newcolors[:1, :] = white\n",
    "    modified_cmap = mcolors.ListedColormap(newcolors)\n",
    "    return modified_cmap\n",
    "\n",
    "def plot_replication_data_vs_data(data, labels, colors, xmin=1e-15, xmax=1e5, ymin=0, ymax=500, title_x=\"Error\", title_y=\"Replication time (min)\",\n",
    "                                  log_x=True, log_y=False, use_density=True, dpi=100, map_to_white=False, saveQ=False):\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=(10, 6))\n",
    "    modified_cmap = create_modified_cmap(map_to_white)\n",
    "    \n",
    "    if use_density:\n",
    "        for (data1, data2), label, color in zip(data, labels, colors):\n",
    "            # Filter out NaN values\n",
    "            mask = ~np.isnan(data2) & ~np.isnan(data1)\n",
    "            data2 = np.array(data2)[mask]\n",
    "            data1 = np.array(data1)[mask]\n",
    "\n",
    "            density = ScatterDensityArtist(ax, data2, data1, cmap=modified_cmap, dpi=dpi)\n",
    "            ax.add_artist(density)\n",
    "\n",
    "        cbar = plt.colorbar(density, ax=ax)\n",
    "        cbar.ax.set_ylabel('Density')\n",
    "    else:\n",
    "        for (data1, data2), label, color in zip(data, labels, colors):\n",
    "            ax.scatter(data2, data1, s=0.02 if label == 'Whole-genome' else 0.1, color=color, label=label)\n",
    "        # Custom legend\n",
    "        handles, labels = ax.get_legend_handles_labels()\n",
    "        new_handles = [plt.Line2D([], [], color=handle.get_facecolor()[0], marker='o', linestyle='', markersize=3) for handle in handles]\n",
    "        ax.legend(handles=new_handles, labels=labels)\n",
    "\n",
    "    # Set x-axis scale to log if log_x is True\n",
    "    if log_x:\n",
    "        ax.set_xscale('log')\n",
    "\n",
    "    # Set y-axis scale to log if log_y is True\n",
    "    if log_y:\n",
    "        ax.set_yscale('log')\n",
    "        ax.set_ylim((ymin, ymax))\n",
    "    else:\n",
    "        ax.set_ylim((ymin, ymax))\n",
    "        ax.set_ylim(ax.get_ylim()[::-1])  # Invert y-axis only if not log scale\n",
    "\n",
    "    ax.set_xlim((xmin, xmax))\n",
    "    ax.set_aspect(aspect='auto')\n",
    "\n",
    "    ax.set_title('')\n",
    "    ax.set_xlabel(title_x)\n",
    "    ax.set_ylabel(title_y)\n",
    "\n",
    "    if saveQ:\n",
    "        plt.savefig('figures/fig_scatter_density.pdf', bbox_inches='tight', transparent=True)\n",
    "    \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "cc665cd4-59ad-41ab-a941-b33d5b2b3241",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_plot(load_type1, load_type2, cell_line, chr_numbers, chr_numbers_tc, chr_numbers_cfs, site_letters,\n",
    "                  show_whole_genome=True, show_telomeres=False, show_centromeres=False, \n",
    "                  show_fragile_sites=True, allcfsQ=False, xmin=1e-15, xmax=1e-3, ymin=0, ymax=500, log_x=True, log_y=False,\n",
    "                  use_density=True, dpi=100, map_to_white=False, saveQ=False):\n",
    "\n",
    "    # Lists to be mapped\n",
    "    list1 = [\"time_data\", \"time_sim\", \"error\", \"fire_rates\", \"forkd\"]\n",
    "    list2 = [\"Replication time (min)\", \"Replication time (min)\", \"Error\", \"Firing rate\", \"Fork directionality\"]\n",
    "    \n",
    "    # Create a dictionary using a dictionary comprehension\n",
    "    title_map = {key: value for key, value in zip(list1, list2)}\n",
    "    \n",
    "    title_x = title_map[load_type2]\n",
    "    title_y = title_map[load_type1]\n",
    "\n",
    "    all_data = []\n",
    "    labels = []\n",
    "    colours = []\n",
    "\n",
    "    if show_whole_genome:\n",
    "        all_data.append(generate_all_data(cell_line, chr_numbers, load_type1, load_type2))\n",
    "        labels.append(\"Whole-genome\")\n",
    "        colours.append('#1f77b4')\n",
    "\n",
    "    if False:\n",
    "        data_t, data_c = generate_telomeres_centromeres_data(cell_line, chr_numbers_tc, load_type1, load_type2)\n",
    "        all_data.append(data_t)\n",
    "        labels.append(\"Telomeres\")\n",
    "        colours.append('orange')\n",
    "\n",
    "    if show_telomeres:\n",
    "        positions_telomeres = np.loadtxt(f'data/genome_regions/telomeres/positions_telomeres_chr[{chr_number}].txt', dtype=int)\n",
    "        all_data.append(positions_telomeres)\n",
    "        labels.append(\"Telomeres\")\n",
    "        colours.append('orange')\n",
    "\n",
    "\n",
    "    if show_centromeres:\n",
    "        data_t, data_c = generate_telomeres_centromeres_data(cell_line, chr_numbers_tc, load_type1, load_type2)\n",
    "        all_data.append(data_c)\n",
    "        labels.append(\"Centromeres\")\n",
    "        colours.append('green')\n",
    "\n",
    "    if show_fragile_sites:\n",
    "        all_data_cfs = generate_fragile_sites_data(cell_line, chr_numbers_cfs, site_letters, load_type1, load_type2, allcfsQ)\n",
    "        all_data.extend(all_data_cfs)\n",
    "        labels.extend([\"Fragile sites\"] if allcfsQ else [f\"FRA{chr_number}{site_letter}\" for chr_number in chr_numbers_cfs for site_letter in site_letters])\n",
    "        colours.extend(list(plt.cm.autumn(np.linspace(0, .5, len(labels) - len(colours)))))\n",
    "\n",
    "    plot_replication_data_vs_data(all_data, labels, colours, xmin, xmax, ymin, ymax, title_x=title_x, title_y=title_y, log_x=log_x, log_y=log_y,\n",
    "                                  use_density=use_density, dpi=dpi, map_to_white=map_to_white, saveQ=saveQ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "56b77076-8f28-4042-9bcf-a80008a3dad9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_plot(load_type1, load_type2, cell_line, chr_numbers,\n",
    "                  show_telomeres=False, show_centromeres=False, show_fragile_sites=False,\n",
    "                  chr_number_fragile_sites=[1], site_letters=['A'],\n",
    "                  xmin=1e-15, xmax=1e-3, ymin=0, ymax=500, log_x=True, log_y=False,\n",
    "                  use_density=True, dpi=100, map_to_white=False, saveQ=False):\n",
    "\n",
    "    global all_data\n",
    "\n",
    "    # Lists to be mapped\n",
    "    list1 = [\"time_data\", \"time_sim\", \"error\", \"fire_rates\", \"forkd\"]\n",
    "    list2 = [\"Replication time (min)\", \"Replication time (min)\", \"Error\", \"Firing rate\", \"Fork directionality\"]\n",
    "    \n",
    "    # Create a dictionary using a dictionary comprehension\n",
    "    title_map = {key: value for key, value in zip(list1, list2)}\n",
    "    \n",
    "    title_x = title_map[load_type2]\n",
    "    title_y = title_map[load_type1]\n",
    "\n",
    "    all_data = []\n",
    "    labels = []\n",
    "    colours = []\n",
    "\n",
    "    all_data.append(generate_all_data(cell_line, chr_numbers, load_type1, load_type2))\n",
    "    labels.append(\"Whole-genome\")\n",
    "    colours.append('#1f77b4')\n",
    "\n",
    "    for chr_number in chr_numbers:\n",
    "\n",
    "        if show_centromeres:\n",
    "            positions_centromeres = load_function_pos(chr_number, \"centromeres\")\n",
    "            data_centromeres = [[sublist[pos] for pos in positions_centromeres] for sublist in all_data[0]]\n",
    "            all_data.append(data_centromeres)\n",
    "            labels.append(\"\" if \"Centromeres\" in labels else \"Centromeres\")\n",
    "            colours.append('green')\n",
    "    \n",
    "        if show_telomeres:\n",
    "            positions_telomeres = load_function_pos(chr_number, \"telomeres\")\n",
    "            data_telomeres = [[sublist[pos] for pos in positions_telomeres] for sublist in all_data[0]]\n",
    "            all_data.append(data_telomeres)\n",
    "            labels.append(\"\" if \"Telomeres\" in labels else \"Telomeres\")\n",
    "            colours.append('orange')\n",
    "\n",
    "    for chr_number in chr_number_fragile_sites:\n",
    "\n",
    "        if show_fragile_sites:\n",
    "            for site_letter in site_letters:\n",
    "                positions_fragile_sites = load_function_pos(chr_number, \"fragile_sites\", site_letter=site_letter)\n",
    "                data_fragile_sites = [[sublist[pos] for pos in positions_fragile_sites] for sublist in all_data[0]]\n",
    "                all_data.append(data_fragile_sites)\n",
    "                labels.append(f'FRA{chr_number}{site_letter}')\n",
    "                colours.extend(list(plt.cm.autumn(np.linspace(0, .5, len(labels) - len(colours)))))\n",
    "\n",
    "    plot_replication_data_vs_data(all_data, labels, colours, xmin, xmax, ymin, ymax, title_x=title_x, title_y=title_y, log_x=log_x, log_y=log_y,\n",
    "                                  use_density=use_density, dpi=dpi, map_to_white=map_to_white, saveQ=saveQ)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cc484f3-f311-4778-84e2-0f858ec76102",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Firing rate plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "e045657b-7d67-4f12-9e4a-3f5051f533e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fire_plotf(cell_line, chr_numbers, resolution, file_name, saveQ, ax=None, aspect_ratio=(10, 6), replace_missing_with_nan=True, show_ticks=True, show_title=True):\n",
    "\n",
    "    # Chromosome lengths in kb (1 kb resolution)\n",
    "    chr_lengths = [249251, 243200, 198023, 191155, 180916, 171116, 159139, 146365, 141214, 135535, 135007, 133852, 115170, 107350, 102532, 90355, 81196, 78078, 59129, 63026, 48130, 51305]\n",
    "\n",
    "    if ax is None:\n",
    "        fig, axes = plt.subplots(len(chr_numbers), 1, figsize=aspect_ratio, sharey=True)\n",
    "    else:\n",
    "        axes = [ax]  # Ensure axes is a list even if it's just one subplot\n",
    "    \n",
    "    for idx, chr_number in enumerate(chr_numbers):\n",
    "        # Data loading: Read firing rates from a text file\n",
    "        firing_rates_file_path = f'data/whole-genome_firing_rates/fire_rates_{cell_line}_chr[{chr_number}].txt'\n",
    "        firing_rates = np.loadtxt(firing_rates_file_path, dtype=float)\n",
    "\n",
    "        if replace_missing_with_nan:\n",
    "            missing_data_path = f'data/whole-genome_missing_data/missing_data_{cell_line}_chr[{chr_number}].txt'\n",
    "            missing_positions = np.loadtxt(missing_data_path, dtype=int)\n",
    "            firing_rates[missing_positions] = np.nan\n",
    "\n",
    "        # Generate chromosome positions in Mb\n",
    "        x = np.linspace(0, chr_lengths[chr_number - 1] / 1000, len(firing_rates))  # Chromosome positions in Mb\n",
    "\n",
    "        # Plotting\n",
    "        ax = axes[idx] if len(chr_numbers) > 1 else axes[0]\n",
    "        ax.plot(x, firing_rates, color='#1f77b4', linewidth=2, alpha=.9)\n",
    "        \n",
    "        if idx == 0 and show_title:\n",
    "            ax.set_title(f'{cell_line}')\n",
    "        if idx == len(chr_numbers) - 1:\n",
    "            if show_ticks:\n",
    "                ax.set_xlabel('Chromosome position (Mb)')\n",
    "                ax.set_xticks(np.arange(0, chr_lengths[0] / 1000, 20))  # Set x-axis ticks to show every 20 Mb\n",
    "        else:\n",
    "            ax.tick_params(axis='x', which='both', bottom=False, top=False, labelbottom=False)  # Remove x-axis ticks and labels for all but the last plot\n",
    "        \n",
    "        ax.set_yscale('log')\n",
    "        ax.set_ylim(10**-12, 10**-3)\n",
    "        ax.set_xlim(0, chr_lengths[0] / 1000)  # Set x-axis limit to the largest chromosome length in Mb\n",
    "        ax.tick_params(axis='both', which='both', direction='out')\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(True)\n",
    "        \n",
    "        # Add chromosome number on the top right corner with a transparent box\n",
    "        ax.text(0.995, 0.93, f'chr {chr_number}', transform=ax.transAxes, \n",
    "                fontsize=12, verticalalignment='top', horizontalalignment='right',\n",
    "                bbox=dict(facecolor='white', alpha=0., edgecolor='none'))\n",
    "        \n",
    "        ax.grid(False)  # Remove grid for each plot\n",
    "\n",
    "    # Add a shared y-axis label if no axis is provided\n",
    "    if ax is None:\n",
    "\n",
    "        fig.text(0.065, 0.5, 'Firing rate', va='center', rotation='vertical')\n",
    "\n",
    "        # Adjust layout\n",
    "        plt.subplots_adjust(left=0.1, right=0.95, top=0.95, bottom=0.05, hspace=0.15)  # Adjust margins and spacing\n",
    "\n",
    "    # Save plot if saveQ is True\n",
    "    if saveQ:\n",
    "        plt.savefig(f'figures/fig_plot_fire_signatures_{cell_line}.svg', bbox_inches='tight', transparent=True)\n",
    "\n",
    "        plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ab3000a4-f985-4f62-bae0-577d56f148cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fire_plotf2(cell_line, chr_number, chrpos_min, chrpos_max, file_name, saveQ, ax=None, show_ticks=True, show_title=True):\n",
    "    global fire_data\n",
    "\n",
    "    # Data loading\n",
    "    fire_data = np.loadtxt('data/whole-genome_firing_rates/fire_rates_' + cell_line + '_chr[' + str(chr_number) + '].txt', dtype=float)[chrpos_min:chrpos_max]\n",
    "    x = np.linspace(chrpos_min, chrpos_max, chrpos_max - chrpos_min)  # Chromosome positions\n",
    "\n",
    "    # Plotting\n",
    "    if ax is None:\n",
    "        plt.figure(figsize=(10, 6))\n",
    "        ax = plt.gca()\n",
    "        \n",
    "    ax.plot(x, fire_data, color='#1f77b4', linewidth=2, alpha=.9)\n",
    "    if show_title:\n",
    "        ax.set_title(cell_line + ' - Chromosome ' + str(chr_number))\n",
    "    ax.set_xlabel('Chromosome position (kb)' if show_title else None)\n",
    "    ax.set_ylabel('Firing rate' if show_title else \"Firing\")\n",
    "    ax.set_yscale('log')\n",
    "    ax.set_ylim(10**-12, 10**-3)\n",
    "    ax.set_xlim(chrpos_min, chrpos_max)  # Ensure the x-axis covers the full range\n",
    "    ax.tick_params(axis='both', which='both', direction='out', bottom=show_ticks, labelbottom=show_ticks, left=show_ticks, labelleft=show_ticks)\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_visible(True)\n",
    "        \n",
    "    # Set logarithmic ticks for y-axis\n",
    "    ax.yaxis.set_major_formatter(LogFormatterMathtext())\n",
    "    ax.yaxis.set_minor_formatter(LogFormatterMathtext())\n",
    "\n",
    "    # Save plot\n",
    "    if saveQ:\n",
    "        plt.savefig('figures/plot_fire_' + file_name + '.pdf', bbox_inches='tight', transparent=True)\n",
    "\n",
    "    if ax is None:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5717de5e-f5fd-4851-9bc0-79c50dc3f51c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Error plots ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "a7bdb482-003b-421c-aefd-5ba6b49617ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "def error_plotf(cell_line, chr_number, chrpos_min, chrpos_max, file_name, saveQ, ax=None, show_ticks=True, show_title=True):\n",
    "    global error_data\n",
    "\n",
    "    # Data loading\n",
    "    error_data = np.loadtxt('data/whole-genome_error/error_' + cell_line + '_chr[' + str(chr_number) + '].txt', dtype=float)[chrpos_min:chrpos_max]\n",
    "    x = np.linspace(chrpos_min, chrpos_max, chrpos_max - chrpos_min)  # Chromosome positions\n",
    "\n",
    "    # Plotting\n",
    "    if ax is None:\n",
    "        plt.figure(figsize=(10, 6))\n",
    "        ax = plt.gca()\n",
    "        \n",
    "    ax.plot(x, error_data, color='#ff7f0e', linewidth=2, alpha=.9)\n",
    "    if show_title:\n",
    "        ax.set_title(cell_line + ' - Chromosome ' + str(chr_number))\n",
    "    ax.set_xlabel('Chromosome position (kb)' if show_title else None)\n",
    "    ax.set_ylabel('Error' if show_title else \"Error\")\n",
    "    ax.set_yscale('log')\n",
    "    ax.set_ylim(10**-12, 10**6)\n",
    "    ax.set_xlim(chrpos_min, chrpos_max)  # Ensure the x-axis covers the full range\n",
    "    ax.tick_params(axis='both', which='both', direction='out', bottom=show_ticks, labelbottom=show_ticks, left=show_ticks, labelleft=show_ticks)\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_visible(True)\n",
    "        \n",
    "    # Set logarithmic ticks for y-axis\n",
    "    ax.yaxis.set_major_formatter(LogFormatterMathtext())\n",
    "    ax.yaxis.set_minor_formatter(LogFormatterMathtext())\n",
    "\n",
    "    # Save plot\n",
    "    if saveQ:\n",
    "        plt.savefig('figures/plot_error_' + file_name + '.pdf', bbox_inches='tight', transparent=True)\n",
    "\n",
    "    if ax is None:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "bb9d79fa-cadc-4b17-9619-f1bfea5981a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_goodness_of_fit(chr_number, saveQ, ax=None, cell_lines=None, chrpos_min=0, chrpos_max=10, alld=True, base_path='data/', shift_param=1.0):\n",
    "    global spaced_data, missing_data\n",
    "    \n",
    "    if cell_lines is None:\n",
    "        cell_lines_BigWig = [\"HeLa-S3\",\"BJ1\",\"IMR90\",\"HUVEC\",\"K562\",\"GM12878\",\"HepG2\",\"MCF-7\"]\n",
    "        cell_lines_HighRes = [\"H1\",\"H9\",\"HCT\"]\n",
    "        cell_lines = cell_lines_BigWig + cell_lines_HighRes\n",
    "\n",
    "    chr_lengths = [249251, 243200, 198023, 191155, 180916, 171116, 159139, 146365, 141214, 135535, 135007, 133852, 115170, 107350, 102532, 90355, 81196, 78078, 59129, 63026, 48130, 51305]\n",
    "    num_positions = chr_lengths[chr_number - 1]\n",
    "    num_cell_lines = len(cell_lines)\n",
    "\n",
    "    # Placeholder for goodness of fit data (use actual data from your model)\n",
    "    goodness_of_fit = [load_function(cell_line, chr_number, 'error') for cell_line in cell_lines]\n",
    "\n",
    "    # Apply shifted log transformation\n",
    "    transformed_goodness_of_fit = [np.log1p(data + shift_param) for data in goodness_of_fit]\n",
    "\n",
    "    # Normalize the transformed data to [0, 1] range\n",
    "    min_val = np.nanmin(transformed_goodness_of_fit)\n",
    "    max_val = np.nanmax(transformed_goodness_of_fit)\n",
    "    normalized_goodness_of_fit = [(data - min_val) / (max_val - min_val) for data in transformed_goodness_of_fit]\n",
    "\n",
    "    # Create a custom colormap for vivid red to vivid green\n",
    "    cmap = plt.get_cmap('RdYlGn_r')  # Note the '_r' to reverse the colormap\n",
    "\n",
    "    # Create a colormap that includes gray for the \"no data\" regions\n",
    "    colors = cmap(np.linspace(0, 1, 256))\n",
    "    colormap_gray = np.array([[0., 0., 0., 0.17]])  # RGBA for gray in colormap\n",
    "    new_colors = np.vstack((colors, colormap_gray))\n",
    "    extended_cmap = ListedColormap(new_colors)\n",
    "\n",
    "    if not alld:\n",
    "        normalized_goodness_of_fit = [i[chrpos_min:chrpos_max] for i in normalized_goodness_of_fit]\n",
    "        num_positions = chrpos_max - chrpos_min\n",
    "        x = np.linspace(chrpos_min, chrpos_max, chrpos_max - chrpos_min)  # Chromosome positions\n",
    "\n",
    "    # Create the spaced data array with triplicates and NaN rows\n",
    "    spaced_data = np.full((num_cell_lines * 4 - 1, num_positions), np.nan)\n",
    "    for i in range(num_cell_lines):\n",
    "        spaced_data[i * 4] = normalized_goodness_of_fit[i]\n",
    "        spaced_data[i * 4 + 1] = normalized_goodness_of_fit[i]\n",
    "        spaced_data[i * 4 + 2] = normalized_goodness_of_fit[i]\n",
    "\n",
    "    # Load missing data and update goodness_of_fit\n",
    "    for i, cell_line in enumerate(cell_lines):\n",
    "        missing_data = load_missing_data(cell_line, chr_number) \n",
    "        if not alld:\n",
    "            #missing_data = list(set(missing_data) & set(range(chrpos_min, chrpos_max)))\n",
    "            missing_data = [i - chrpos_min for i in range(chrpos_min, chrpos_max) if i in missing_data]\n",
    "        spaced_data[i * 4, missing_data] = 256  # Mark missing data positions with index for gray\n",
    "        spaced_data[i * 4 + 1, missing_data] = 256\n",
    "        spaced_data[i * 4 + 2, missing_data] = 256\n",
    "\n",
    "    # Avoiding the maximum value being exactly 1 by subtracting a small epsilon value\n",
    "    spaced_data0 = spaced_data\n",
    "    spaced_data = np.clip(spaced_data * 255, 0, 255 - 1)  # Normalize to 0-255 range\n",
    "    spaced_data[spaced_data0 == 256] = 256  # Ensure missing data stays at 256\n",
    "\n",
    "    # Normalize data for colormap\n",
    "    if ax is None:\n",
    "        fig, ax = plt.subplots(figsize=(15, .5*num_cell_lines))\n",
    "        \n",
    "    cax = ax.imshow(spaced_data, aspect='auto', cmap=extended_cmap, interpolation='nearest', vmin=0, vmax=255)\n",
    "\n",
    "    # Add color bar to the right and match height of the bars on the left\n",
    "    if alld:\n",
    "        cbar = fig.colorbar(cax, orientation='vertical', pad=0.02)\n",
    "        cbar.set_label('Normalized error' if alld else None)\n",
    "        cbar.set_ticks(np.linspace(0, 255, 6))\n",
    "        cbar.set_ticklabels(np.round(np.linspace(0, 1, 6), 2))\n",
    "\n",
    "    # Set ticks and labels\n",
    "    yticks_positions = np.arange(1, num_cell_lines * 4, 4)\n",
    "    ax.set_yticks(yticks_positions if alld else [])\n",
    "    ax.set_yticklabels(cell_lines if alld else [])\n",
    "    xtick_positions = np.arange(0, num_positions, 20000)\n",
    "    xtick_labels = (xtick_positions / 1000).astype(int)\n",
    "    ax.set_xticks(xtick_positions if alld else [])\n",
    "    ax.set_xticklabels(xtick_labels if alld else [])\n",
    "    ax.set_xlabel('Chromosome position (Mb)' if alld else None)\n",
    "\n",
    "    ax.grid(False)\n",
    "\n",
    "    # Add a gray square and text for \"Missing data\" in the top right corner\n",
    "    if alld:\n",
    "        ax.text(0, 1.04, f'Chromosome {chr_number}', transform=ax.transAxes, fontsize=7, verticalalignment='top', horizontalalignment='left', bbox=dict(facecolor=\"none\", edgecolor='none'))\n",
    "        ax.text(1, 1.04, 'Missing data', transform=ax.transAxes, fontsize=7, verticalalignment='top', horizontalalignment='right', bbox=dict(facecolor=\"none\", edgecolor='none'))\n",
    "        ax.add_patch(plt.Rectangle((.924, 1.02), 0.008, 0.02, transform=ax.transAxes, color=[0., 0., 0., 0.17], clip_on=False))\n",
    "\n",
    "    if saveQ:\n",
    "        plt.savefig(f'figures/fig_goodness_of_fit_chr[{chr_number}].pdf', bbox_inches='tight', transparent=True)\n",
    "\n",
    "    if ax is None:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "109ff973-cce2-440f-8ffb-7dc73e1a66db",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Fork directionality plots ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "c1810da7-ce5e-407c-bc46-6dde078d8b60",
   "metadata": {},
   "outputs": [],
   "source": [
    "def forkd_plotf(cell_line, chr_number, chrpos_min, chrpos_max, file_name, spec_fileQ, saveQ, ax=None, show_ticks=True, show_title=True):\n",
    "    global time_data\n",
    "\n",
    "    # Data loading (Warning: requires saving data in fitting procedure)\n",
    "    # Choose between whole-genome files or particular simulation\n",
    "    if spec_fileQ:\n",
    "        forkd_data = np.loadtxt(f'data/whole-genome_fork_directionality/fork_directionality_{cell_line}_chr[{chr_number}]_{chrpos_min}-{chrpos_max}.txt', dtype=float)\n",
    "    else:\n",
    "        forkd_data = np.loadtxt(f'data/whole-genome_fork_directionality/fork_directionality_{cell_line}_chr[{chr_number}].txt', dtype=float)[chrpos_min:chrpos_max]\n",
    "    x = np.linspace(chrpos_min, chrpos_max, chrpos_max - chrpos_min)  # Chromosome positions\n",
    "\n",
    "    \n",
    "    # Plotting\n",
    "    if ax is None:\n",
    "        plt.figure(figsize=(10, 6))\n",
    "        ax = plt.gca()\n",
    "        \n",
    "    ax.plot(x, forkd_data, label='', color='black', linewidth=2, alpha=0.6)\n",
    "    if show_title:\n",
    "        ax.set_title(cell_line + ' - Chromosome ' + str(chr_number))\n",
    "    ax.set_xlabel('Chromosome position (kb)' if show_title else None)\n",
    "    ax.set_ylabel('Fork directionality' if show_title else \"Fork dir.\")\n",
    "    ax.set_ylim(-1, 1)\n",
    "    ax.set_xlim(chrpos_min, chrpos_max)  # Ensure the x-axis covers the full range\n",
    "    #ax.legend(loc='lower right')<s\n",
    "    ax.grid(True)\n",
    "    ax.grid(False)\n",
    "    ax.tick_params(axis='both', which='both', direction='out', bottom=show_ticks, labelbottom=show_ticks, left=show_ticks, labelleft=show_ticks)\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_visible(True)\n",
    "    \n",
    "    # Save plot\n",
    "    if saveQ:\n",
    "        plt.savefig('figures/plot_forkd_' + file_name + '.pdf', bbox_inches='tight', transparent=True)\n",
    "\n",
    "    if ax is None:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e58b9eda-0fd0-43e2-ab1d-3d4f15038a75",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Ideogram plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "2cbdfb19-d0cb-4b73-a9e2-35fc802f3361",
   "metadata": {},
   "outputs": [],
   "source": [
    "def show_genome(cell_line, chr_number, chrpos_min, chrpos_max,\n",
    "                show_genes_allQ=True, show_genes_bandsQ=True, show_genesQ=False,\n",
    "                show_rt_plotQ=True, show_fire_plotQ=True, show_forkd_plotQ=False, show_error_plotQ=True, show_error_heat_plotQ=True,\n",
    "                show_axisQ=True):\n",
    "    chrom = f\"chr{chr_number}\"\n",
    "    start = chrpos_min * 1000\n",
    "    end = chrpos_max * 1000\n",
    "\n",
    "    chr_lengths = [249251, 243200, 198023, 191155, 180916, 171116, 159139, 146365, 141214, 135535, 135007, 133852, 115170, 107350, 102532, 90355, 81196, 78078, 59129, 63026, 48130, 51305]\n",
    "\n",
    "    fig = plt.figure(figsize=(10, 10), dpi=150)\n",
    "\n",
    "    fig.suptitle(f\"{cell_line} - Chromosome {chr_number}\")\n",
    "\n",
    "    height_ratios0 = [0.1, 0.1, 0.05, .2, .4, .2, .2, .2, .1, 0., 1]\n",
    "    show_components = [show_genes_allQ, show_genes_allQ, show_genes_bandsQ, show_genesQ, show_rt_plotQ, show_fire_plotQ, show_forkd_plotQ, show_error_plotQ, show_error_heat_plotQ, show_axisQ, True]\n",
    "    height_ratios = [height_ratios0[i] for i in range(len(height_ratios0)) if show_components[i]]\n",
    "\n",
    "    gs_i = -1\n",
    "\n",
    "    gs = fig.add_gridspec(\n",
    "        nrows=len(height_ratios),\n",
    "        ncols=3,\n",
    "        width_ratios=[2, 20, 2],\n",
    "        height_ratios=height_ratios,\n",
    "        hspace=0.01,\n",
    "        left=0.01,\n",
    "        right=0.99,\n",
    "        top=0.94,\n",
    "    )\n",
    "\n",
    "    # Add the full chromosome ideogram\n",
    "    if show_genes_allQ:\n",
    "        gs_i += 1\n",
    "        all_chrom_ax = fig.add_subplot(gs[gs_i, 1])\n",
    "        all_chrom_ax.axis(\"off\")\n",
    "        all_chrom_ax.set_xticks([])\n",
    "        pyideogram.ideogramh(chrom, ax=all_chrom_ax)\n",
    "        \n",
    "        # Draw a red rectangle on top of the region depicted in the full chromosome ideogram\n",
    "        rect_start = start  # Start position in base pairs\n",
    "        rect_end = end  # End position in base pairs\n",
    "        rectangle_thickness = 3  # Set the thickness of the rectangle outline\n",
    "        all_chrom_ax.add_patch(plt.Rectangle((rect_start, -0.5), rect_end - rect_start, 1, edgecolor='red', facecolor='none', linewidth=rectangle_thickness, clip_on=False))\n",
    "        \n",
    "        # Draw lines from the bottom corners of the rectangle\n",
    "        #all_chrom_ax.plot([rect_start, -chr_lengths[chr_number-1] * 1e3 * (1/5.25)], [-0.5, -2], color='black', linewidth=1, linestyle='-', clip_on=False)\n",
    "        #all_chrom_ax.plot([rect_end, chr_lengths[chr_number-1] * 1e3+chr_lengths[chr_number-1] * 1e3 * (1/5.3)], [-0.5, -2], color='black', linewidth=1, linestyle='-', clip_on=False)\n",
    "\n",
    "        gs_i += 1\n",
    "        empty_ax = fig.add_subplot(gs[gs_i, :])\n",
    "        empty_ax.axis(\"off\")\n",
    "        empty_ax.tick_params(labelbottom=False)\n",
    "        #empty_ax.plot([rect_start, -chr_lengths[chr_number-1] * 1e3 * (1/5.25)], [-0.5, -2], color='black', linewidth=1, linestyle='-', clip_on=False)\n",
    "\n",
    "    # Add the zoomed ideogram within the same interval\n",
    "    if show_genes_bandsQ:\n",
    "        gs_i += 1\n",
    "        interval_ideogram_ax = fig.add_subplot(gs[gs_i, :])\n",
    "        interval_ideogram_ax.axis(\"off\")\n",
    "        interval_ideogram_ax.set_xlim(start, end)\n",
    "        interval_ideogram_ax.set_xticks([])\n",
    "        pyideogram.ideogramh(chrom, ax=interval_ideogram_ax, names=True)\n",
    "        zoom_ax = interval_ideogram_ax\n",
    "\n",
    "    # Add the gene track plot\n",
    "    if show_genesQ:\n",
    "        gs_i += 1\n",
    "        gene_track_ax = fig.add_subplot(gs[gs_i, :])\n",
    "        pyideogram.genetrack(\n",
    "            f\"{chrom}:{start}-{end}\",\n",
    "            ax=gene_track_ax,\n",
    "            textlane=True,\n",
    "            transcriptstyle=\"arrowed\",\n",
    "            exonstyle=\"Box\",\n",
    "        )\n",
    "        # Set genome ticks on the gene track plot\n",
    "        pyideogram.set_genome_xticks(gene_track_ax)\n",
    "        # Remove axis labels and ticks for the gene track plot\n",
    "        gene_track_ax.axis(\"on\")\n",
    "        gene_track_ax.set_xticks([])\n",
    "        gene_track_ax.set_ylabel('Genes')\n",
    "        gene_track_ax.tick_params(labelbottom=False)\n",
    "        if not show_genes_bandsQ:\n",
    "            zoom_ax = gene_track_ax\n",
    "\n",
    "    # Add RT plot\n",
    "    if show_rt_plotQ:\n",
    "        gs_i += 1\n",
    "        rt_ax = fig.add_subplot(gs[gs_i, :])\n",
    "        rt_plotf(cell_line, chr_number, chrpos_min, chrpos_max, 6, 'example_file', False, False, ax=rt_ax, show_ticks=False, show_title=False)\n",
    "\n",
    "    # Add firing plot\n",
    "    if show_fire_plotQ:\n",
    "        gs_i += 1\n",
    "        fire_ax = fig.add_subplot(gs[gs_i, :])\n",
    "        fire_plotf2(cell_line, chr_number, chrpos_min, chrpos_max, 'example_file', False, ax=fire_ax, show_ticks=False, show_title=False)\n",
    "\n",
    "    # Add fork directionality plot\n",
    "    if show_forkd_plotQ:\n",
    "        gs_i += 1\n",
    "        forkd_ax = fig.add_subplot(gs[gs_i, :])\n",
    "        forkd_plotf(cell_line, chr_number, chrpos_min, chrpos_max, 'example_file', False, False, ax=forkd_ax, show_ticks=False, show_title=False)\n",
    "\n",
    "    # Add the error plot\n",
    "    if show_error_plotQ:\n",
    "        gs_i += 1\n",
    "        error_ax = fig.add_subplot(gs[gs_i, :])\n",
    "        error_plotf(cell_line, chr_number, chrpos_min, chrpos_max, 'example_file', False, ax=error_ax, show_ticks=False, show_title=False)\n",
    "    \n",
    "    # Add the error heat plot\n",
    "    if show_error_heat_plotQ:\n",
    "        gs_i += 1\n",
    "        error_heat_ax = fig.add_subplot(gs[gs_i, :])\n",
    "        error_heat_ax.set_ylabel('Error')\n",
    "        plot_goodness_of_fit(chr_number, False, ax=error_heat_ax, cell_lines=[cell_line], chrpos_min=chrpos_min, chrpos_max=chrpos_max, alld=False)\n",
    "\n",
    "    # Add a new row showing just the x-axis with the ticks\n",
    "    if show_axisQ:\n",
    "        gs_i += 1\n",
    "        x_ticks_ax = fig.add_subplot(gs[gs_i, :])\n",
    "        x_ticks_ax.set_xlim(start, end)\n",
    "        x_ticks_ax.xaxis.set_visible(True)\n",
    "        x_ticks_ax.spines['top'].set_visible(False)\n",
    "        x_ticks_ax.spines['left'].set_visible(False)\n",
    "        x_ticks_ax.spines['right'].set_visible(False)\n",
    "        x_ticks_ax.tick_params(axis='x', which='both', bottom=True, top=False, labelbottom=True)\n",
    "        # Hide y-axis and labels\n",
    "        x_ticks_ax.tick_params(left=False, labelleft=False)\n",
    "        x_ticks_ax.set_yticks([])\n",
    "        # Set a dummy plot for ticks visibility\n",
    "        x_ticks_ax.plot([start, end], [0, 0], color='white', alpha=0)  # Ensure the axis is plotted correctly\n",
    "    \n",
    "        # Determine tick step and format\n",
    "        tick_step = (end - start) // 5\n",
    "        if tick_step < 1:\n",
    "            tick_step = 1\n",
    "        tick_labels = []\n",
    "        tick_positions = []\n",
    "        for x in range(start, end + tick_step, tick_step):\n",
    "            tick_positions.append(x)\n",
    "            if (end - start) < 10000000:  # Less than 10 Mb range\n",
    "                tick_labels.append(f'{x / 10**6:.1f}'.rstrip('0').rstrip('.'))\n",
    "            else:\n",
    "                tick_labels.append(f'{x / 10**6:.0f}')\n",
    "        \n",
    "        x_ticks_ax.set_xticks(tick_positions)\n",
    "        x_ticks_ax.set_xticklabels(tick_labels)\n",
    "        \n",
    "        # Add a label to the x-axis\n",
    "        x_ticks_ax.set_xlabel('Chromosome position (Mb)')\n",
    "        \n",
    "        # Add extra ticks\n",
    "        minor_tick_step = tick_step // 5\n",
    "        if minor_tick_step > 0:\n",
    "            minor_ticks = []\n",
    "            for x in range(start, end, minor_tick_step):\n",
    "                if x not in tick_positions:\n",
    "                    minor_ticks.append(x)\n",
    "            x_ticks_ax.set_xticks(minor_ticks, minor=True)\n",
    "            x_ticks_ax.tick_params(axis='x', which='minor', bottom=True, top=False, labelbottom=False)\n",
    "\n",
    "\n",
    "\n",
    "    # Zoom the ideogram\n",
    "    if show_genes_allQ:\n",
    "        pyideogram.zoom(zoom_ax, all_chrom_ax)\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6abe46bd-c113-4a81-88c3-465153938b22",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
